File: .codevalid/tests/task_5124983147_20260211210101/component_events_flash_messages.py
import pytest
from bs4 import BeautifulSoup

# Simulate the ERB logic for flash messages in app/views/layouts/application.html.erb.
def render_flash_messages(notice=None, alert=None):
    html = ""
    if notice:
        html += f'<div class="notice"><span class="flash-message">{notice}</span><button class="dismiss-notice">x</button></div>'
    if alert:
        html += f'<div class="alert"><span class="flash-message">{alert}</span><button class="dismiss-alert">x</button></div>'
    return html

def get_soup(html):
    return BeautifulSoup(html, "html.parser")

@pytest.mark.describe("Flash Messages (Notice/Alert) Component")
class TestFlashMessages:

    @pytest.mark.it("Render notice message when notice is present")
    def test_render_notice_message_when_notice_is_present(self):
        # Given
        notice = "Event created successfully"
        alert = None
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        # Then
        notice_div = soup.find("div", class_="notice")
        assert notice_div is not None
        assert "Event created successfully" in notice_div.text
        alert_div = soup.find("div", class_="alert")
        assert alert_div is None

    @pytest.mark.it("Render alert message when alert is present")
    def test_render_alert_message_when_alert_is_present(self):
        # Given
        notice = None
        alert = "Event could not be created"
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        # Then
        alert_div = soup.find("div", class_="alert")
        assert alert_div is not None
        assert "Event could not be created" in alert_div.text
        notice_div = soup.find("div", class_="notice")
        assert notice_div is None

    @pytest.mark.it("Render both notice and alert messages when both are present")
    def test_render_both_notice_and_alert_messages_when_both_are_present(self):
        # Given
        notice = "Event updated"
        alert = "Some fields missing"
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        # Then
        notice_div = soup.find("div", class_="notice")
        alert_div = soup.find("div", class_="alert")
        assert notice_div is not None
        assert alert_div is not None
        assert "Event updated" in notice_div.text
        assert "Some fields missing" in alert_div.text

    @pytest.mark.it("Do not render any message when neither notice nor alert is present")
    def test_do_not_render_any_message_when_neither_notice_nor_alert_is_present(self):
        # Given
        notice = None
        alert = None
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        # Then
        assert soup.find("div", class_="notice") is None
        assert soup.find("div", class_="alert") is None

        # Also test with empty strings
        html_empty = render_flash_messages(notice="", alert="")
        soup_empty = get_soup(html_empty)
        assert soup_empty.find("div", class_="notice") is None
        assert soup_empty.find("div", class_="alert") is None

    @pytest.mark.it("Do not render notice message when notice is empty string")
    def test_do_not_render_notice_message_when_notice_is_empty_string(self):
        # Given
        notice = ""
        alert = None
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        # Then
        assert soup.find("div", class_="notice") is None

    @pytest.mark.it("Do not render alert message when alert is empty string")
    def test_do_not_render_alert_message_when_alert_is_empty_string(self):
        # Given
        notice = None
        alert = ""
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        # Then
        assert soup.find("div", class_="alert") is None

    @pytest.mark.it("Clear notice message after user action")
    def test_clear_notice_message_after_user_action(self):
        # Given
        notice = "Event deleted"
        alert = None
        # When (initial render)
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        notice_div = soup.find("div", class_="notice")
        assert notice_div is not None
        # When (user clicks dismiss)
        html_after_dismiss = render_flash_messages(notice=None, alert=alert)
        soup_after = get_soup(html_after_dismiss)
        # Then
        assert soup_after.find("div", class_="notice") is None

    @pytest.mark.it("Clear alert message after user action")
    def test_clear_alert_message_after_user_action(self):
        # Given
        notice = None
        alert = "Access denied"
        # When (initial render)
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        alert_div = soup.find("div", class_="alert")
        assert alert_div is not None
        # When (user clicks dismiss)
        html_after_dismiss = render_flash_messages(notice=notice, alert=None)
        soup_after = get_soup(html_after_dismiss)
        # Then
        assert soup_after.find("div", class_="alert") is None

    @pytest.mark.it("Render notice message with long text")
    def test_render_notice_message_with_long_text(self):
        # Given
        long_text = "T" * 255
        notice = long_text
        alert = None
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        notice_div = soup.find("div", class_="notice")
        # Then
        assert notice_div is not None
        assert long_text in notice_div.text
        assert len(notice_div.text) >= 255

    @pytest.mark.it("Render alert message with special characters")
    def test_render_alert_message_with_special_characters(self):
        # Given
        alert = '<script>alert("xss")</script> & ©'
        notice = None
        # When
        html = render_flash_messages(notice=notice, alert=alert)
        soup = get_soup(html)
        alert_div = soup.find("div", class_="alert")
        # Then
        assert alert_div is not None
        # Should display as plain text, not executed as HTML/JS
        assert '<script>' in alert_div.text
        assert '&' in alert_div.text
        assert '©' in alert_div.text
        # Ensure the HTML is not interpreted (i.e., no <script> tag in parsed children)
        assert alert_div.find("script") is None

File: .codevalid/tests/task_5124983147_20260211210101/component_events_flash_messages.rb
require 'test_helper'
require 'action_view/testing/resolvers'

class ComponentEventsFlashMessagesTest < ActionView::TestCase
  tests ActionView::Base

  def setup
    @view = ActionView::Base.new(ActionView::LookupContext.new([]), {}, nil)
    @template = File.read('app/views/layouts/application.html.erb')
  end

  def render_with_flash(notice: nil, alert: nil)
    @view.assign(flash: { notice: notice, alert: alert })
    @view.render(inline: @template)
  end

  def test_render_notice_message_when_notice_is_present
    html = render_with_flash(notice: 'Event created successfully')
    assert_includes html, 'Event created successfully'
    assert_includes html, 'class="notice"'
  end

  def test_render_alert_message_when_alert_is_present
    html = render_with_flash(alert: 'Event could not be created')
    assert_includes html, 'Event could not be created'
    assert_includes html, 'class="alert"'
  end

  def test_render_both_notice_and_alert_messages_when_both_are_present
    html = render_with_flash(notice: 'Event updated', alert: 'Some fields missing')
    assert_includes html, 'Event updated'
    assert_includes html, 'Some fields missing'
  end

  def test_do_not_render_any_message_when_neither_notice_nor_alert_is_present
    html = render_with_flash
    refute_includes html, 'notice'
    refute_includes html, 'alert'
  end

  def test_do_not_render_notice_message_when_notice_is_empty_string
    html = render_with_flash(notice: '')
    refute_includes html, 'class="notice"'
  end

  def test_do_not_render_alert_message_when_alert_is_empty_string
    html = render_with_flash(alert: '')
    refute_includes html, 'class="alert"'
  end

  def test_render_notice_message_with_long_text
    long_text = 'T' * 255
    html = render_with_flash(notice: long_text)
    assert_includes html, long_text
  end

  def test_render_alert_message_with_special_characters
    msg = '<script>alert("xss")</script> & ©'
    html = render_with_flash(alert: msg)
    assert_includes html, CGI.escapeHTML(msg)
  end

  def test_render_nothing_when_notice_and_alert_are_whitespace_only
    html = render_with_flash(notice: '   ', alert: '   ')
    refute_includes html, 'class="notice"'
    refute_includes html, 'class="alert"'
  end

  def test_notice_styling_uses_correct_css_class
    html = render_with_flash(notice: 'Event saved')
    assert_includes html, 'class="notice"'
  end

  def test_alert_styling_uses_correct_css_class
    html = render_with_flash(alert: 'Failed to save')
    assert_includes html, 'class="alert"'
  end

  def test_order_of_messages_when_both_notice_and_alert_displayed
    html = render_with_flash(notice: 'A', alert: 'B')
    assert html.index('A') < html.index('B')
  end

  def test_notice_message_persists_across_redirect
    get root_path, params: {}, flash: { notice: 'Persisted' }
    follow_redirect! if response.redirect?
    assert_includes response.body, 'Persisted'
  end

  def test_alert_message_persists_across_redirect
    get root_path, params: {}, flash: { alert: 'Persisted alert' }
    follow_redirect! if response.redirect?
    assert_includes response.body, 'Persisted alert'
  end

  def test_flash_messages_do_not_expose_html_injection_via_interpolation
    html = render_with_flash(notice: 'Error <b>bold</b>')
    assert_includes html, 'Error &lt;b&gt;bold&lt;/b&gt;'
  end

  def test_render_flash_messages_at_top_of_main_content_area
    html = render_with_flash(notice: 'Saved')
    first_occurrence = html.index('Saved')
    assert first_occurrence < 200
  end

  def test_render_only_latest_flash_messages_on_redirect
    get root_path, params: {}, flash: { notice: 'Old message' }
    follow_redirect! if response.redirect?
    get root_path, params: {}, flash: { notice: 'New message' }
    follow_redirect! if response.redirect?
    assert_includes response.body, 'New message'
    refute_includes response.body, 'Old message'
  end

  def test_do_not_render_non_string_flash_values
    html = render_with_flash(notice: { a: 1 })
    refute_includes html, 'class="notice"'
  end
end
